Remove sandbox extension to fonts service

pvollan opened the pull request:
#### d853f70a64ac5696c3b102a20fa02802ee3394bb
<pre>
Remove sandbox extension to fonts service
<a href="https://bugs.webkit.org/show_bug.cgi?id=284233">https://bugs.webkit.org/show_bug.cgi?id=284233</a>
<a href="https://rdar.apple.com/138204781">rdar://138204781</a>

Reviewed by Sihui Liu.

On macOS, we currently allow access to the fonts service when the app has enabled user installed fonts.
To support blocking of the font service on macOS in the WebContent process, we need to register the URLs
of user installed fonts in the WebContent process, so that CoreText is aware of them. In the UI process,
we are enumerating the user installed fonts, and then send the URLs along with a sandbox extension to
the font file to the WebContent process, where they are registered.

This patch also fixes an exception we made for the Books app on iOS when blocking the Mobile asset
service. To support removing the exception for Books, we have to register some fonts in the WebContent
process. The URLs for the fonts are gathered in the UI process, and sent to the WebContent process along
with a sandbox extension, so they can be registered.

Additionally, this patch guards the enablement of MobileGestalt and Mobile asset blocking on the define
ENABLE(REMOVE_XPC_AND_MACH_SANDBOX_EXTENSIONS_IN_WEBCONTENT).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::additionalFonts):
(WebKit::WebProcessPool::registerUserInstalledFonts):
(WebKit::WebProcessPool::registerAssetFont):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
(WebKit::WebProcessProxy::platformInitialize):
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerAdditionalFonts):

Canonical link: <a href="https://commits.webkit.org/288425@main">https://commits.webkit.org/288425@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/293ee0e32eafdac861adc20b20594910d8857c12

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83105 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2746 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37400 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88214 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34151 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2822 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10669 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64707 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22468 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86160 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2071 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75573 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44987 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/82537 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1978 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29773 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33185 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/76089 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73141 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30473 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89580 "Hash 293ee0e3 for PR 37598 does not build (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/82150 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7506 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/89580 "Hash 293ee0e3 for PR 37598 does not build (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10618 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71385 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/89580 "Hash 293ee0e3 for PR 37598 does not build (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16548 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15288 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12863 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10341 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15846 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104558 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10212 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25306 "Found 3 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/substring-global-atom-cache.js.lockdown, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11981 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+{
+    AdditionalFonts additionalFonts;
+    for (auto& fontURL : fontURLs) {
+        auto sandboxExtensionHandle = SandboxExtension::createHandleWithoutResolvingPath(fontURL.fileSystemPath(), SandboxExtension::Type::ReadOnly);
```
`if (auto sandboxExtensionHandle = SandboxExtension::createHandleWithoutResolvingPath(fontURL.fileSystemPath(), SandboxExtension::Type::ReadOnly))`

pvollan replied:
Thanks for reviewing! I will address these issues in the upcoming patch.

cdumez replied:
Ideally, this whole thing would be written using `WTF::compactMap()`:
```
additionalFonts.fontDataList = WTF::compactMap(fontURLs, [](auto& fontURL) -> std::optional<FontData> {
    if (auto sandboxExtensionHandle = SandboxExtension::createHandleWithoutResolvingPath(fontURL.fileSystemPath(), SandboxExtension::Type::ReadOnly))
        return FontData { fontURL, WTFMove(*sandboxExtensionHandle) };
    return std::nullopt;
});
```

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+void WebProcessPool::registerUserInstalledFonts(WebProcessProxy& process)
+{
+    auto blockPtr = makeBlockPtr([weakThis = WeakPtr { *this }] {
+        NSString *userInstalledFontsPath = [NSHomeDirectory() stringByAppendingPathComponent:@"Library/Fonts"];
```
`RetainPtr`

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+            }
+        }
+
+        RunLoop::protectedMain()->dispatch([weakThis, fontURLs = WTFMove(fontURLs)] {
```
`crossThreadCopy(WTFMove(fontURLs))`

cdumez replied:
`weakThis = WTFMove(weakThis)` too.

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+
+        RunLoop::protectedMain()->dispatch([weakThis, fontURLs = WTFMove(fontURLs)] {
+            if (RefPtr protectedThis = weakThis.get()) {
+                for (auto& process : protectedThis->m_processes)
```
`Ref process`?

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+    if (m_additionalFontURLs) {
+        process.send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(*m_additionalFontURLs)), 0);
+        return;
+    }
```
We should create `blockPtr` after this early return, or it's unused

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+        process.send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(*m_additionalFontURLs)), 0);
+        return;
+    }
+    static bool hasStartedCachingUserInstalledFonts = false;
```
hmm shouldn't this state `hasStartedCachingUserInstalledFonts` be specific to the `WebProcessPool`?

cdumez replied:
I think Sihui is likely right here.

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+
+void WebProcessPool::registerAssetFont(ASCIILiteral fontName, WebProcessProxy& process)
+{
+    auto nsFontName = fontName.createNSString();
```
`RetainPtr`

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+{
+    auto nsFontName = fontName.createNSString();
+    NSDictionary *attributes = @{ (NSString *)kCTFontFamilyNameAttribute: nsFontName.get(), (NSString *)kCTFontRegistrationScopeAttribute: @(kCTFontPriorityComputer) };
+    auto description = adoptCF(CTFontDescriptorCreateWithAttributes((__bridge CFDictionaryRef)attributes));
```
`RetainPtr`

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+    NSArray *descriptions = [NSArray arrayWithObject:(__bridge id)description.get()];
+
+    auto blockPtr = makeBlockPtr([nsFontName, weakThis = WeakPtr { *this }](CTFontDescriptorMatchingState state, CFDictionaryRef progressParameter) {
+        if (state == kCTFontDescriptorMatchingDidFinish) {
```
Might early return here to reduce nested statements:
```
if (state != kCTFontDescriptorMatchingDidFinish)
    return true;
RELEASE_LOG(Process, "Font matching finished, progress parameter = %@", (__bridge id)progressParameter);
...
```

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                    auto cfFontURL = adoptCF(static_cast<CFURLRef>(CTFontCopyAttribute(font.get(), kCTFontURLAttribute)));
+                    URL fontURL(cfFontURL.get());
+                    RELEASE_LOG(Process, "Registering font name %@ with url %s", nsFontName.get(), fontURL.string().utf8().data());
+                    for (auto& process : protectedThis->m_processes)
```
`Ref process`?

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+    }
+
+    static bool hasStartedCachingAssetFonts = false;
+    if (!hasStartedCachingAssetFonts) {
```
Same question here: why is this shared?

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+        return true;
+    });
+
+    if (m_additionalFontURLs) {
```
This does not seem right: `WebProcessPool::registerAssetFont` takes `fontName`, which means caller can register a font that's not in `m_additionalFontURLs`. Shouldn't we check if the requested font is in `m_additionalFontURLs` first and return early only if the font already exists?

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                    if (protectedThis->m_additionalFontURLs)
+                        protectedThis->m_additionalFontURLs->append(fontURL);
+                    else
+                        protectedThis->m_additionalFontURLs = { fontURL };
```
We might write it as:
```
if (!protectedThis->m_additionalFontURLs)
    protectedThis->m_additionalFontURLs = Vector<URL> { };
protectedThis->m_additionalFontURLs.append(WTFMove(fontURL));
```

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+        RunLoop::protectedMain()->dispatch([weakThis, fontURLs = WTFMove(fontURLs)] {
+            if (RefPtr protectedThis = weakThis.get()) {
+                for (auto& process : protectedThis->m_processes)
+                    process->send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(fontURLs)), 0);
```
I' guessing this is right but wanted to double check with you. Are we intentionally re-creating new SandboxExtensions for the same files for each WebProcess? I'm guessing these things get consumed and cannot be re-used for several processes but wanted to check.

pvollan replied:
Yes, good catch, this is intended behavior. I haven't added an audit token to target specific WebContent processes, but will do that in the next patch.

Thanks for reviewing! I will address the comments in the upcoming patch.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+    auto description = adoptCF(CTFontDescriptorCreateWithAttributes((__bridge CFDictionaryRef)attributes));
+    NSArray *descriptions = [NSArray arrayWithObject:(__bridge id)description.get()];
+
+    auto blockPtr = makeBlockPtr([nsFontName, weakThis = WeakPtr { *this }](CTFontDescriptorMatchingState state, CFDictionaryRef progressParameter) {
```
`nsFontName = WTFMove(nsFontName)`

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+    auto blockPtr = makeBlockPtr([nsFontName, weakThis = WeakPtr { *this }](CTFontDescriptorMatchingState state, CFDictionaryRef progressParameter) {
+        if (state == kCTFontDescriptorMatchingDidFinish) {
+            RELEASE_LOG(Process, "Font matching finished, progress parameter = %@", (__bridge id)progressParameter);
+            RunLoop::protectedMain()->dispatch([weakThis, nsFontName] {
```
We should `WTFMove()` the captured variables here.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+            RELEASE_LOG(Process, "Font matching finished, progress parameter = %@", (__bridge id)progressParameter);
+            RunLoop::protectedMain()->dispatch([weakThis, nsFontName] {
+                if (RefPtr protectedThis = weakThis.get()) {
+                    auto font = adoptCF(CTFontCreateWithName((__bridge CFStringRef)nsFontName.get(), 0.0, nullptr));
```
`RetainPtr`

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+            RunLoop::protectedMain()->dispatch([weakThis, nsFontName] {
+                if (RefPtr protectedThis = weakThis.get()) {
+                    auto font = adoptCF(CTFontCreateWithName((__bridge CFStringRef)nsFontName.get(), 0.0, nullptr));
+                    auto cfFontURL = adoptCF(static_cast<CFURLRef>(CTFontCopyAttribute(font.get(), kCTFontURLAttribute)));
```
`RetainPtr`

cdumez replied:
Actually, if it is only used to construct a `WTF::URL`, we should just construct the `WTF::URL` right away and avoid the local cf variable.

cdumez commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -66,6 +66,7 @@
 #endif
 
 #if PLATFORM(COCOA)
+#include "AdditionalFonts.h"
```
Can we forward declare instead?

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1534,6 +1534,23 @@ static float currentBacklightLevel()
+        SandboxExtension::consumePermanently(fontData.sandboxExtensionHandle);
+
+        RetainPtr<CFURLRef> cfURL = fontData.fontURL.createCFURL();
+        NSArray* fontURLs = @[(__bridge NSURL *)cfURL.get()];
```
Use RetainPtr here and WTFMove() one the next line.

szewai commented on Source/WebKit/Shared/AdditionalFonts.h:
```diff
@@ -0,0 +1,47 @@
+};
+
+struct AdditionalFonts {
+    Vector<FontData> fontDataList;
```
Why do we want to use a `struct` instead of directly using `Vector<FontData>`?

pvollan replied:
I created a separate `struct` for this, since I think it is likely that we will add additional members to this struct in the future, but let me know if you think we should avoid this, and I'll change it to `Vector<FontData>` :)

Thanks for reviewing!

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,100 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+
+        if (sandboxExtensionHandle)
+            return FontData { fontURL, WTFMove(*sandboxExtensionHandle) };
+        return FontData { fontURL, SandboxExtension::Handle() };
```
If we cannot create sandbox extension for the handle, do we still need to send it to web process?

pvollan replied:
Yes, we currently have cases where the WebContent process has access to the font file, but the UI process does not. In these cases, we can still successfully register the font in the WebContent process without the extension.

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,100 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+        process.send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(*m_userInstalledFontURLs, process.auditToken())), 0);
+        return;
+    }
+
```
We can move m_hasStartedCachingUserInstalledFonts check up here:
```
// Process will be notified after fetching fonts.
if (m_hasStartedCachingUserInstalledFonts)
    return;

m_hasStartedCachingUserInstalledFonts = true;
```

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,100 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+
+        RunLoop::protectedMain()->dispatch([weakThis = WTFMove(weakThis), fontURLs = crossThreadCopy(WTFMove(fontURLs))] {
+            if (RefPtr protectedThis = weakThis.get()) {
+                for (Ref process : protectedThis->m_processes)
```
Technically I think only the processes that are invoked with this function should get notified?

pvollan replied:
That is a very good point. In the latest patch, we only send the message to the WebContent process invoked with this function. This change also enables the removal of the `m_hasStartedCachingUserInstalledFonts` flag.

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,100 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+            if ([font hasSuffix:@"ttf"] || [font hasSuffix:@"ttc"] || [font hasSuffix:@"otf"]) {
+                RetainPtr fontsPath = [userInstalledFontsPath stringByAppendingPathComponent:font];
+                auto fontURL = URL::fileURLWithFileSystemPath(String(fontsPath.get()));
+                fontURLs.append(fontURL);
```
`WTFMove(fontURL)`

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1333,4 +1333,100 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+
+static URL fontURLFromName(NSString *nsFontName)
+{
+    RetainPtr font = adoptCF(CTFontCreateWithName((__bridge CFStringRef)nsFontName, 0.0, nullptr));
```
`ASCIILiteral` has `createCFString()`, so this could take `ASCIILiteral` directly

szewai commented on Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
```diff
@@ -71,6 +73,14 @@
+        protectedProcessPool()->registerAssetFont("Canela Text"_s, *this);
+        protectedProcessPool()->registerAssetFont("Proxima Nova"_s, *this);
+        protectedProcessPool()->registerAssetFont("Publico Text"_s, *this);
+    }
```
My understanding is only web process that loads page with preference `blockMobileAssetInWebContentSandbox()` true can get the font, but here we are giving it to all processes. Not sure if this is expected behavior.

Also with existing usage it seems  `registerAssetFont` does not need to take font -- it can decide the list by itself: like if isIBooks(), the list contains 3 items, else it's empty; similar to what we do for `registerUserInstalledFonts`.

pvollan replied:
That is a good point! Since the preference `blockMobileAssetInWebContentSandbox()` is only a temporary preference meant for staging, and not a preference the client should be modifying, I think it is OK to set this for all processes. I expect the preference to be removed fairly soon. However, let me know if you think we should avoid this!

pvollan committed:
Remove sandbox extension to fonts service
https://bugs.webkit.org/show_bug.cgi?id=284233
rdar://138204781

Reviewed by Sihui Liu.

On macOS, we currently allow access to the fonts service when the app has enabled user installed fonts.
To support blocking of the font service on macOS in the WebContent process, we need to register the URLs
of user installed fonts in the WebContent process, so that CoreText is aware of them. In the UI process,
we are enumerating the user installed fonts, and then send the URLs along with a sandbox extension to
the font file to the WebContent process, where they are registered.

This patch also fixes an exception we made for the Books app on iOS when blocking the Mobile asset
service. To support removing the exception for Books, we have to register some fonts in the WebContent
process. The URLs for the fonts are gathered in the UI process, and sent to the WebContent process along
with a sandbox extension, so they can be registered.

Additionally, this patch guards the enablement of MobileGestalt and Mobile asset blocking on the define
ENABLE(REMOVE_XPC_AND_MACH_SANDBOX_EXTENSIONS_IN_WEBCONTENT).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::additionalFonts):
(WebKit::WebProcessPool::registerUserInstalledFonts):
(WebKit::WebProcessPool::registerAssetFont):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
(WebKit::WebProcessProxy::platformInitialize):
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerAdditionalFonts):

Canonical link: https://commits.webkit.org/288425@main
